### PR TITLE
Add `vault_quota_rate_limit` resource

### LINKED ISSUE #824: [feature request] Support Resource Quota
Vault 1.5 adds support for [rate limit quota](https://www.vaultproject.io/api/system/rate-limit-quotas). 

This is a feature request to support resource quota.

(I can work on this when I have the time, but raising this issue so that someone else might want to take it up first.)

### PR DESCRIPTION
**Pull request by lawliet89**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #824

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
add `vault_quota_rate_limit` resource to manage resource quota limit
```

Output from acceptance testing:

```
$ VAULT_ADDR='http://127.0.0.1:8200' VAULT_TOKEN=12345 make testacc TESTARGS='-run=TestQuotaRateLimit'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestQuotaRateLimit -timeout 120m
?       github.com/terraform-providers/terraform-provider-vault [no test files]
?       github.com/terraform-providers/terraform-provider-vault/cmd/coverage    [no test files]
?       github.com/terraform-providers/terraform-provider-vault/cmd/generate    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/codegen (cached) [no tests to run]
?       github.com/terraform-providers/terraform-provider-vault/generated       [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/datasources/transform/decode  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/datasources/transform/encode  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/alphabet  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/role      (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/template  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/transformation    (cached) [no tests to run]
?       github.com/terraform-providers/terraform-provider-vault/schema  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/util    (cached) [no tests to run]
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.42s)
PASS
ok      github.com/terraform-providers/terraform-provider-vault/vault   (cached)

...
```

### CODE REVIEW

**Comment by lawliet89** [comment]
Test failure is unrelated:

```
=== RUN   TestAccIdentityGroupMemberEntityIdsNonExclusive

--- FAIL: TestAccIdentityGroupMemberEntityIdsNonExclusive (0.48s)

    testing.go:640: Step 2 error: Check failed: unexpected member entity id 98b9b473-e3de-fd10-16f5-1cd8266bfa07 in group 8583f549-15c7-a891-6d1a-147c76fcec72

```

**Comment by lawliet89** [comment]
@catsby Sorry for the silence. Let me look into your comments ASAP.

File: website/docs/r/quota_rate_limit.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# vault\_pki\_secret\_backend\_intermediate\_cert\_request
+
+Generates a new private key and a CSR for signing the PKI Secret Backend.
+
+~> **Important** All data provided in the resource configuration will be
+written in cleartext to state and plan files generated by Terraform, and
+will appear in the console output when Terraform runs. Protect these
+artifacts accordingly. See
+[the main provider documentation](../index.html)
+for more details.
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply website/docs/r/quota_rate_limit.md]: This is embarrassing. I forgot to go back and fix this. Now fixed.

File: vault/resource_quota_rate_limit.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+			"rate": {
+				Type:        schema.TypeFloat,
+				Required:    true,
+				Description: "The maximum number of requests at any given second to be allowed by the quota rule. The rate must be positive.",
+			},
+		},
+	}
+}
+
+func quotaRateLimitUpdateFields(d *schema.ResourceData, data map[string]interface{}) {
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply vault/resource_quota_rate_limit.go:47]: Removed.

File: vault/resource_quota_rate_limit.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+			},
+			"path": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Path of the mount or namespace to apply the quota. A blank path configures a global rate limit quota.",
+			},
+			"rate": {
+				Type:        schema.TypeFloat,
+				Required:    true,
+				Description: "The maximum number of requests at any given second to be allowed by the quota rule. The rate must be positive.",
```
[the remark this answers is not in the document]
Comment (lawliet89) [reply vault/resource_quota_rate_limit.go:41]: Added.

**Comment by lawliet89** [comment]
@catsby I have made the changes you suggested.

I cannot replicate your test failure. What version of Vault are you running the tests against? The test passes on CI and it's running against the `latest` tag and I am running against `1.5.3`.

File: website/docs/r/quota_rate_limit.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (catsby) [reply website/docs/r/quota_rate_limit.md]: 👍

**Comment by catsby** [comment]
Hey @lawliet89 - I've been running the tests with a Vault server compiled from the `master` branch of Vault, which could be part of my issues. I did try with `1.5.3` and got the test to pass, however I still got an error if I ran with `count=2`:

```
[terraform-provider-vault][quota-rate-limit](4)$ VAULT_TOKEN=root make testacc TEST=./vault/... TESTARGS="-run=TestQuota -count=2" 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./vault/... -v -run=TestQuota -count=2 -timeout 120m
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.21s)
=== RUN   TestQuotaRateLimit
    testing.go:640: Step 0 error: errors during apply:

        Error: Error creating Resource Rate Limit Quota tf-test-2918301973751561055: Error making API request.

        URL: PUT http://127.0.0.1:8200/v1/sys/quotas/rate-limit/tf-test-2918301973751561055
        Code: 400. Errors:

        * quota rule with similar properties exists under the name "tf-test-9053389438201676050"

          on /var/folders/9s/yrf0vw2j2xbfgktx9x16tk6c0000gp/T/tf-test001805183/main.tf line 2:
          (source code not available)


--- FAIL: TestQuotaRateLimit (0.06s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-vault/vault   0.547s
FAIL
make: *** [testacc] Error 1
```

Can you try running that command? I'm running Vault 1.5.3 with `vault server -dev -dev-root-token-id=root`

**Comment by lawliet89** [comment]
Vault doesn't like it when you have multiple quotas for the same path. 

I can add another separate test to support running the tests in parallel for different secret mounts but I also do want to have a test for `""`. The test for `""` cannot be run simultaneously it seems. What do you think?



```

**Comment by catsby** [comment]
Well, I was assuming that after the test completed the first time, all quotas would be removed before the second test ran. That would mean any previous quota at that path would have been deleted, so there shouldn't be a conflict. Am I misunderstanding something here?

**Comment by lawliet89** [comment]
I think both tests were running simultaneously and not one after another.

**Comment by lawliet89** [comment]
I modified the tests and made the attribute `path` be `ForceNew`. That seemed to have fixed the issues.

```
$ VAULT_ADDR='http://127.0.0.1:8200' VAULT_TOKEN=12345 make testacc TESTARGS="-run=TestQuota -count 10"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestQuota -count 10 -timeout 120m
?       github.com/terraform-providers/terraform-provider-vault [no test files]
?       github.com/terraform-providers/terraform-provider-vault/cmd/coverage    [no test files]
?       github.com/terraform-providers/terraform-provider-vault/cmd/generate    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/codegen 0.296s [no tests to run]
?       github.com/terraform-providers/terraform-provider-vault/generated       [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/datasources/transform/decode  0.825s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/datasources/transform/encode  0.542s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/alphabet  0.484s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/role      0.549s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/template  0.265s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/transformation    0.419s [no tests to run]
?       github.com/terraform-providers/terraform-provider-vault/schema  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/util    0.229s [no tests to run]
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.50s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.47s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.46s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.48s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.43s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.44s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.92s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.56s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.78s)
=== RUN   TestQuotaRateLimit
--- PASS: TestQuotaRateLimit (0.46s)
PASS
ok      github.com/terraform-providers/terraform-provider-vault/vault   5.530s
```

**Review by catsby — APPROVED** [review]
LGTM thanks!

**Pull request merged**